File: 0948-bag-of-tokens/0948-bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end());
        int left = 0;
        int right = tokens.size() - 1;
        int currScore = 0;
        int maxScore = 0;
        while(left <= right) {
            if(tokens[left] <= power) {
                power -= tokens[left];
                currScore += 1;
                left += 1;
            } else {
                if(currScore >= 1) {
                    power += tokens[right];
                    currScore -= 1;
                }
                right -= 1;
            }
            maxScore = max(maxScore, currScore);
        }
        return maxScore;
    }
};

File: 0948-bag-of-tokens/0948-bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int left = 0;
        int right = tokens.length - 1;
        int currScore = 0;
        int maxScore = 0;
        
        while (left <= right) {
            if (tokens[left] <= power) {
                power -= tokens[left];
                currScore++;
                left++;
            } else {
                if (currScore >= 1) {
                    power += tokens[right];
                    currScore--;
                }
                right--;
            }
            maxScore = Math.max(maxScore, currScore);
        }
        
        return maxScore;
    }
}
